### ISSUE TITLE
DSL.val(T) could be improved to support converter discovery for subtypes/supertypes.

### ISSUE THREAD

**Issue by jezovuk** [jOOQ/jOOQ#7473]
### Expected behavior and actual behavior:

Expecting `DSL.val(T)` to use registered converter for superclass.

Instead, it crashes:

```
org.jooq.exception.SQLDialectNotSupportedException: Type class java.time.ZoneRegion is not supported in dialect DEFAULT
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:856)
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:800)
	at org.jooq.impl.DSL.getDataType(DSL.java:20527)
	at org.jooq.impl.DSL.val(DSL.java:18331)
        . . .
```

### Steps to reproduce the problem:

Forced type with converter for converting `java.time.ZoneId` to PostgreSQL `varchar`:

```
forcedType {
    userType = 'java.time.ZoneId'
    converter = 'xena.common.util.jooq.ZoneIdConverter'
    expression = '.*(?:tz|zone)'
    types = 'varchar'
}
```

Converter:

```
package xena.common.util.jooq;

import java.time.ZoneId;

import org.jooq.impl.AbstractConverter;

/**
 * jOOQ converter for converting {@link String} to {@link ZoneId} and vice versa.
 */
@SuppressWarnings("serial")
public final class ZoneIdConverter extends AbstractConverter<String, ZoneId> {

    public ZoneIdConverter() {
        super(String.class, ZoneId.class);
    }

    @Override
    public ZoneId from(String str) {
        return str != null ? ZoneId.of(str) : null;
    }

    @Override
    public String to(ZoneId zone) {
        return zone != null ? zone.getId() : null;
    }

}

```

Using `DSL.val(T)` to pass ZoneId value to the query, like this:

```
DSL.val(ZoneId.of("Europe/Belgrade"))
```

### Versions:

- jOOQ: 3.10.6
- Java: 1.8.152
- Database (include vendor): PostgreSQL 9.6.5
- OS: Windows 10
- JDBC Driver (include name if inofficial driver): pg jdbc 42.2.2

### Analysis:

It seems that the culprit is `org.jooq.impl.DefaultDataType.getDataType(SQLDialect, Class<T>, DataType<T>)`. There, in line 846 (jOOQ 3.10.6), `DataType` instance is obtained from static map:

```
return (DataType<T>) SQL_DATATYPES_BY_TYPE.get(type);
```

If I'm not mistaken, `SQL_DATATYPES_BY_TYPE` is populated upon `DefaultDataType` instance creation (which includes data types with custom converters). In my use case (with `ZoneId`), the map will contain entry with `ZoneId.class` as key and corresponding data type (with my converter) as value.

However, in example above, `ZoneId.of("Europe/Belgrade")` returns an instance of `java.time.ZoneRegion`. ZoneRegion is subtype of ZoneId, but `ZoneId.class` is not equal to `ZoneRegion.class`. Hence, data type is not found in the map and converter is not used. Instead, an exception occurs.

I do not have enough understanding of jOOQ internals to suggest a complete solution.
However, it seems that (at least for some use cases) data type detection code can be relaxed to use `Class.isAssignableFrom()` for matching parameter type with map keys instead of `Map.get()` (i.e. `Object.equals()`). Direction of `isAssignableFrom()` check (key-to-param or param-to-key) should be determined by calling context and passed to `org.jooq.impl.DefaultDataType.getDataType(SQLDialect, Class<T>, DataType<T>)` as param (depending how we want to use the type - convert from DB to Java or convert from Java value to DB). In other words, we could get data type for subclass (if exact one is not available) when conversion from database value to Java type is needed and, conversely, get data type for superclass when conversion from Java value to database type is needed (as in my zone-region example). Obviously, in some cases neither will do (in which case throwing an exception is fine). With suggested change, it is possible that there are multiple subclass/superclass data types registered (matching param type); this could either also lead to an exception or some kind of deterministic tie-breaking algorithm could be used.

Again, I cannot estimate full impact of such change, so treat this only as a suggestion / enhancement idea, not an actual bug/issue.

Obvious workaround is to use `DSL.val(Object, DataType<T>)`. Example:

```
DSL.val(ZoneId.of("Europe/Belgrade"), DSL.getDataType(ZoneId.class))
```

**Comment by lukaseder** [comment]
Thanks for your report. `DSL.val()` cannot be adapted to find out dynamically what the most appropriate converter would have been. Yes, there's a static map in the internals somewhere, but discussing that map isn't going to help, because that's an implementation detail. The question is how the API should be used, and possibly, improved.

Yes, using `DSL.val(Object, DataType)` is the appropriate solution. It's not a workaround, there simply isn't a better way right now. Do note that the fact that you could use `DSL.getDataType(ZoneId.class)` is actually the above implementation detail leaking into the API. The correct usage would be to have an explicit `DataType` reference somewhere in your code:

```java
public static final DataType<ZoneId> DT_ZONE_ID = SQLDataType.VARCHAR.asConvertedDataType(new ZoneIdConverter());
```

And then:

```java
DSL.val(ZoneId.of("Europe/Belgrade"), DT_ZONE_ID)
```

In many cases, the data type can be derived from the context, though. Since you're using a `<forcedType/>`, you'll have columns that are of that data type. So, forming a predicate on such a column will automatically apply the columns data type also on the bind variable:

```java
MY_ZONE.eq(ZoneId.of("Europe/Belgrade")) // Implicit call to DSL.val(.., MY_ZONE.getDataType())
```

I hope this helps

**Comment by lukaseder** [comment]
Note, there's also a pending feature request to allow for a more global type lookup registry: https://github.com/jOOQ/jOOQ/issues/5713

It will not apply to static method calls like `DSL.val()`, though...

**Comment by jezovuk** [comment]
Thanks for answering. I'm aware that what I described is mostly internal implementation stuff, but it seemed unexpected that `val(ZoneId.of("..."))` would not use registered converter for `ZoneId`.

As for creating my own static `DataType` instance instead of using `DSL.getDataType()`, I was doing exactly that before (recently) discovering the latter. I assummed that using public lookup method to find existing data type was the preferred way. Either way is fine.

In any case, this is really a minor issue, so current behavior is quite ok. Thanks again.
